src/ripple/core/impl/ThreadEntry.cpp

```diff
@@ -0,0 +1,66 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/core/ThreadEntry.h>
+
+#include <boost/coroutine/exceptions.hpp>
+
+namespace ripple {
+
+#ifndef NO_LOG_UNHANDLED_EXCEPTIONS
+namespace detail {
+thread_local
+std::string threadName;
+
+void setThreadName(std::string name)
+{
+    threadName = std::move(name);
+}
+}
+
+void terminateHandler()
+{
+    if (std::current_exception())
+    {
+        try
+        {
+            throw;
```

Comment:
That's a good observation Scott.  Details for the interested:

Exception throwing (on gcc and clang) is a 2-phase process.  Phase 1 walks the stack _without_ unwinding it (or executing destructors) looking for a matching catch clause.  When a matching catch clause is found, then phase 2 kicks in and actually unwinds the stack up to the matching catch, executing destructors along the way.  It is the second phase which erases the stack trace (up to the matching catch).  By placing a matching catch at the top of the thread stack, we ensure that we always erase the stack trace up to that top matching catch.

The good part is where phase 1 _doesn't_ find a matching catch.  In that case the compiler calls _down_ the stack to `std::terminate()`.  As `terminate()` is called, the entire stack above the throw is still intact.  `terminal()` then calls the handler (down the stack again).  In the handler we throw and catch all in the same tiny local area of the stack (still under the point of throw).  So the stack only gets unwound from this local `throw;` to this local `catch()`, which is to say, not at all.  After the handler completes, `std::abort()` is called.  And everything above the point of throw (and a little bit of stuff below the throw) is still on the stack, waiting for the abort process to log it.